Sources/GRPC/GRPCKeepaliveHandlers.swift

```diff
@@ -0,0 +1,281 @@
+    self.pingHandler.channelHandler = self
+  }
+
+  /// The connection manager.
+  private var connectionManager: ConnectionManager
+
+  /// The ping handler.
+  private let pingHandler: PingHandler
+
+  func userInboundEventTriggered(context: ChannelHandlerContext, event: Any) {
+    self.pingHandler.userInboundEventTriggered(context: context, event: event)
+    context.fireUserInboundEventTriggered(event)
+  }
+
+  func channelRead(context: ChannelHandlerContext, data: NIOAny) {
+    switch self.unwrapInboundIn(data).payload {
+    case let .ping(pingData, ack: ack):
+      self.pingHandler.read(pingData: pingData, ack: ack, context: context)
+    default:
+      break
+    }
+
+    context.fireChannelRead(data)
+  }
+
+  func send(payload: HTTP2Frame.FramePayload, context: ChannelHandlerContext) {
+    let frame = self.wrapOutboundOut(.init(streamID: .rootStream, payload: payload))
+    _ = context.writeAndFlush(frame)
+  }
+
+  func close(context: ChannelHandlerContext) {
+    connectionManager.idle()
+    context.close(mode: .all, promise: nil)
+  }
+}
+
+internal class GRPCServerKeepaliveHandler: ChannelInboundHandler, ChannelKeepaliveHandler {
+  typealias InboundIn = HTTP2Frame
+  typealias OutboundOut = HTTP2Frame
+
+  init(configuration: ServerConnectionKeepalive) {
+    self.pingHandler = PingHandler(
+      pingCode: 10,
+      time: configuration.time,
+      timeout: configuration.timeout,
+      permitWithoutCalls: configuration.permitWithoutCalls,
+      maximumPingsWithoutData: configuration.maximumPingsWithoutData,
+      minimumSentPingIntervalWithoutData: configuration.minimumSentPingIntervalWithoutData,
+      minimumReceivedPingIntervalWithoutData: configuration.minimumReceivedPingIntervalWithoutData,
+      maximumPingStrikes: configuration.maximumPingStrikes
+    )
+    self.pingHandler.channelHandler = self
+  }
+
+  /// The ping handler.
+  private let pingHandler: PingHandler
+
+  func userInboundEventTriggered(context: ChannelHandlerContext, event: Any) {
+    self.pingHandler.userInboundEventTriggered(context: context, event: event)
+    context.fireUserInboundEventTriggered(event)
+  }
+
+  func channelRead(context: ChannelHandlerContext, data: NIOAny) {
+    switch self.unwrapInboundIn(data).payload {
+    case let .ping(pingData, ack: ack):
+      self.pingHandler.read(pingData: pingData, ack: ack, context: context)
+    default:
+      break
+    }
+
+    context.fireChannelRead(data)
+  }
+
+  func send(payload: HTTP2Frame.FramePayload, context: ChannelHandlerContext) {
+    let frame = self.wrapOutboundOut(.init(streamID: .rootStream, payload: payload))
+    _ = context.writeAndFlush(frame)
+  }
+
+  func close(context: ChannelHandlerContext) {
+    context.close(mode: .all, promise: nil)
+  }
+}
+
+protocol ChannelKeepaliveHandler: AnyObject {
+  func send(payload: HTTP2Frame.FramePayload, context: ChannelHandlerContext)
+  func close(context: ChannelHandlerContext)
+}
+
+private final class PingHandler {
```

Comment:
I think this factoring out of the shared logic is really useful, but I think we should reconsider this exact approach.

The biggest downside of this approach, as noted, is that the keep alive handler needs to hold a reference to the ping handler, and the ping handler needs ta reference to the keep alive handler. This strong cycle needs to be broken. You've done this here using `weak`, which is fine, but it's arguably the worst of the options we have (the other two options are `unowned` which is semantically just fine here and performs better, or even better still to have a direct strong reference that is simply optional, and nil the ref out in `handlerAdded`).

However, an even better design is to flip this class to a struct and drop the reference to the parent object altogether. This object will then be driven by the parent, and will ask the parent to act on its behalf using return values. That is `func read(pingData: HTTP2PingData, ack: Bool, context: ChannelHandlerContext)` becomes `func read(pingData: HTTP2PingData, ack: Bool, context: ChannelHandlerContext) -> Action`, where `Action` is some enum that describes the possible results of this operation:

```swift
struct PingHandler {
    enum Action {
        case cancelScheduledTimeout
        case goawayAndClose
        case reply(HTTP2Frame)
    }
}
```

and so on.

The advantage of this design (and of putting all the state in this struct but all the side effects elsewhere) is twofold. Firstly, it makes testing _way_ easier: you can make the `PingHandler` struct `internal` but prefix it with an underscore, and then you can simply drive it directly in testing, which allows you to easily test any complex logic in isolation.

Secondly, Swift's law of exclusivity will force us into a useful programming pattern. A common danger in NIO-based programs is to only partially perform a state transition before you make outcalls (i.e. calls to other code you don't control). For example, succeeding a promise before you've moved all your state. These outcalls are potentially re-entrant, and so can lead you to nasty state management bugs where your code observes impossible states.

However, because Swift forbids two mutating calls to the same object at the same time, if this ever happens when you are using mutating functions on a struct, either the compiler will catch it or the runtime will force it to crash, depending on the complexity of the pattern. This is really great! Swift will police us and prevent us from making this mistake, which is really nice.

While we're here, you've defined a `ChannelKeepaliveHandler` protocol, which is great. That protocol can be rewritten to be a useful place for code sharing. For example, you could do:

```swift
protocol _ChannelKeepaliveHandler {
    var pingHandler: PingHandler { get set }
}

extension _ChannelKeepaliveHandler {
    func read(pingData: HTTP2PingData, ack: Bool, context: ChannelHandlerContext) {
        let action = self.pingHandler.read(pingData: pingData, ack: ack)
        switch action {
            // Do whatever
        }
    }
}
```

In essence your protocol can define the stored properties necessary to perform the computation, and then define all the computation in protocol extensions to achieve the code re-use. This allows us to really trim down the code we need.